process/elasticproc/tokens/sovereignIndexTokensHandler.go

```diff
@@ -0,0 +1,166 @@
+func NewSovereignIndexTokensHandler(mainChainElastic factory.ElasticConfig, esdtPrefix string) (*sovereignIndexTokensHandler, error) {
+	var mainChainElasticClient elasticproc.DatabaseClientHandler
+	if mainChainElastic.Enabled {
+		var err error
+		argsEsClient := elasticsearch.Config{
+			Addresses:     []string{mainChainElastic.Url},
+			Username:      mainChainElastic.UserName,
+			Password:      mainChainElastic.Password,
+			Logger:        &logging.CustomLogger{},
+			RetryOnStatus: []int{http.StatusConflict},
+			RetryBackoff:  retryBackOff,
+		}
+		mainChainElasticClient, err = client.NewElasticClient(argsEsClient)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &sovereignIndexTokensHandler{
+		indexingEnabled:        mainChainElastic.Enabled,
+		mainChainElasticClient: mainChainElasticClient,
+		esdtPrefix:             esdtPrefix,
+	}, nil
+}
+
+func retryBackOff(attempt int) time.Duration {
+	return time.Duration(math.Exp2(float64(attempt))) * time.Second
+}
+
+// IndexCrossChainTokens will index the new tokens properties
+func (sit *sovereignIndexTokensHandler) IndexCrossChainTokens(elasticClient elasticproc.DatabaseClientHandler, scrs []*data.ScResult, buffSlice *data.BufferSlice) error {
+	if !sit.indexingEnabled {
+		return nil
+	}
+
+	notFoundTokens, err := sit.getTokensFromScrs(elasticClient, scrs)
+	if err != nil {
+		return err
+	}
+
+	if len(notFoundTokens) == 0 { // no new tokens
+		return nil
+	}
+
+	// get tokens from main chain elastic db
+	mainChainTokens := &data.ResponseTokenInfo{}
+	err = sit.mainChainElasticClient.DoMultiGet(context.Background(), notFoundTokens, indexerdata.TokensIndex, true, mainChainTokens)
+	if err != nil {
+		return err
+	}
+
+	return sit.indexNewTokens(mainChainTokens.Docs, buffSlice)
+}
+
+func (sit *sovereignIndexTokensHandler) getTokensFromScrs(elasticClient elasticproc.DatabaseClientHandler, scrs []*data.ScResult) ([]string, error) {
+	receivedTokensIDs := make([]string, 0)
+	for _, scr := range scrs {
+		if scr.SenderShard == core.MainChainShardId {
+			for _, token := range scr.Tokens {
+				tokenPrefix, hasPrefix := esdt.IsValidPrefixedToken(token)
+				if !hasPrefix || tokenPrefix != sit.esdtPrefix {
+					receivedTokensIDs = append(receivedTokensIDs, token)
+				}
+				if isEsdt, tokenCollection := getTokenCollection(hasPrefix, token); isEsdt {
+					receivedTokensIDs = append(receivedTokensIDs, tokenCollection)
+				}
+			}
+		}
+	}
+
+	if len(receivedTokensIDs) == 0 {
+		return make([]string, 0), nil
+	}
+
+	responseTokens := &data.ResponseTokens{}
+	err := elasticClient.DoMultiGet(context.Background(), receivedTokensIDs, indexerdata.TokensIndex, true, responseTokens)
+	if err != nil {
+		return nil, err
+	}
+
+	newTokens := make([]string, 0)
+	for _, token := range responseTokens.Docs {
+		if !token.Found {
+			newTokens = append(newTokens, token.ID)
+		}
+	}
+
+	return newTokens, nil
+}
+
+func getTokenCollection(hasPrefix bool, tokenIdentifier string) (bool, string) {
```

Comment:
Do not return bool + string.
Simply return string and check if len(string) != 0

Reply:
done